File: lua/src/phonenumber_ae.lua
--- Validate, parse and format international phone numbers (ABI v5).
---
--- The idiomatic Lua surface over the phonenumber engine. Carries no
--- phone-number logic — every function here marshals to the C extension in
--- `phonenumber_ae_native` (lua/src/phonenumber_ae.c), which in turn calls the
--- `aether_pn_embed_*` ABI. One engine, one set of behaviours, N language
--- surfaces.
---
---     local pn = require("phonenumber_ae")
---     local num = pn.parse("+1 650 253 0000", "US")
---     print(num:national_number())                       -- "6502530000"
---     print(pn.is_valid_number("US", "+1 650 253 0000"))  -- true
---     print(pn.format("US", "6502530000", pn.INTERNATIONAL)) -- "+1 650-253-0000"
---
---     local ayt = pn.AsYouTypeFormatter.new("US")
---     local last
---     for c in ("6502530000"):gmatch(".") do last = ayt:input_digit(c) end
---     -- last == "(650) 253-0000"
---
---     for _, m in ipairs(pn.find_numbers("call 201-555-0123 today", "US")) do
---       print(m.raw)                                      -- "201-555-0123"
---     end

local native = require("phonenumber_ae_native")

local M = {}

-- ---- ABI constants, re-exported (append only, never renumber) ----

--- Format styles. NOTE: in v2, E164 is 0 (it was 2 in v1).
M.E164          = native.E164
M.INTERNATIONAL = native.INTERNATIONAL
M.NATIONAL      = native.NATIONAL
M.RFC3966       = native.RFC3966

--- Number types (number_type result; -1 = unknown).
M.TYPE_UNKNOWN         = native.TYPE_UNKNOWN
M.TYPE_FIXED_LINE      = native.TYPE_FIXED_LINE
M.TYPE_MOBILE          = native.TYPE_MOBILE
M.TYPE_TOLL_FREE       = native.TYPE_TOLL_FREE
M.TYPE_PREMIUM_RATE    = native.TYPE_PREMIUM_RATE
M.TYPE_SHARED_COST     = native.TYPE_SHARED_COST
M.TYPE_VOIP            = native.TYPE_VOIP
M.TYPE_PERSONAL_NUMBER = native.TYPE_PERSONAL_NUMBER
M.TYPE_PAGER           = native.TYPE_PAGER
M.TYPE_UAN             = native.TYPE_UAN
M.TYPE_VOICEMAIL       = native.TYPE_VOICEMAIL
M.TYPE_FIXED_LINE_OR_MOBILE = native.TYPE_FIXED_LINE_OR_MOBILE

--- ValidationResult (is_possible_number_with_reason).
M.VR_IS_POSSIBLE            = native.VR_IS_POSSIBLE
M.VR_IS_POSSIBLE_LOCAL_ONLY = native.VR_IS_POSSIBLE_LOCAL_ONLY
M.VR_INVALID_COUNTRY_CODE   = native.VR_INVALID_COUNTRY_CODE
M.VR_TOO_SHORT              = native.VR_TOO_SHORT
M.VR_INVALID_LENGTH         = native.VR_INVALID_LENGTH
M.VR_TOO_LONG               = native.VR_TOO_LONG

--- MatchType (is_number_match).
M.MATCH_NOT_A_NUMBER = native.MATCH_NOT_A_NUMBER
M.MATCH_NO_MATCH     = native.MATCH_NO_MATCH
M.MATCH_SHORT_NSN    = native.MATCH_SHORT_NSN
M.MATCH_NSN          = native.MATCH_NSN
M.MATCH_EXACT        = native.MATCH_EXACT

--- CountryCodeSource (pn_source).
M.SRC_FROM_NUMBER_WITH_PLUS    = native.SRC_FROM_NUMBER_WITH_PLUS
M.SRC_FROM_NUMBER_WITH_IDD     = native.SRC_FROM_NUMBER_WITH_IDD
M.SRC_FROM_NUMBER_WITHOUT_PLUS = native.SRC_FROM_NUMBER_WITHOUT_PLUS
M.SRC_FROM_DEFAULT_COUNTRY     = native.SRC_FROM_DEFAULT_COUNTRY

--- Matcher leniency.
M.LENIENCY_POSSIBLE = native.LENIENCY_POSSIBLE
M.LENIENCY_VALID    = native.LENIENCY_VALID

--- ShortNumberCost (short_expected_cost).
M.COST_TOLL_FREE     = native.COST_TOLL_FREE
M.COST_STANDARD_RATE = native.COST_STANDARD_RATE
M.COST_PREMIUM_RATE  = native.COST_PREMIUM_RATE
M.COST_UNKNOWN       = native.COST_UNKNOWN

-- ---- metadata ----

--- The country calling code for a region ("1", "44", …), or "" if unknown.
function M.country_code(region)
  return native.country_code(region)
end

--- An example national number for the region, or "".
function M.example_number(region)
  return native.example_number(region)
end

--- An example national number of a given TYPE_* for the region, or "".
function M.example_number_for_type(region, ntype)
  return native.example_number_for_type(region, ntype)
end

--- An example number that is the right shape but invalid, or "".
function M.invalid_example_number(region)
  return native.invalid_example_number(region)
end

--- The possible-lengths spec for the region (e.g. "9,10"), or "".
function M.possible_lengths(region)
  return native.possible_lengths(region)
end

--- The main region for a country calling code ("44" -> "GB"), or "".
function M.region_code_for_country_code(cc)
  return native.region_code_for_country_code(tostring(cc))
end

--- True if the region is part of the North American Numbering Plan.
function M.is_nanpa_country(region)
  return native.is_nanpa_country(region)
end

--- The national-direct-dial prefix for a region (e.g. "0" or "1"), or "".
--- `strip_non_digits` is a boolean.
function M.ndd_prefix_for_region(region, strip_non_digits)
  return native.ndd_prefix_for_region(region, strip_non_digits and true or false)
end

--- How many region ids the metadata carries.
function M.region_count()
  return native.region_count()
end

--- The region id at a 1-based `index`, or "" when out of range. (The ABI is
--- 0-based; the C extension does the conversion so Lua stays 1-based.)
function M.region_at(index)
  return native.region_at(index)
end

--- Every region id the metadata carries, as a table (engine order).
function M.regions()
  local out = {}
  for i = 1, native.region_count() do
    out[i] = native.region_at(i)
  end
  return out
end

--- The regions, sorted — the deterministic version of `regions()`.
function M.sorted_regions()
  local out = M.regions()
  table.sort(out)
  return out
end

--- Every region that shares a country calling code ("1" -> {"US","CA",…}).
function M.regions_for_country_code(cc)
  cc = tostring(cc)
  local out = {}
  local n = native.cc_region_count(cc)
  for i = 1, n do
    out[i] = native.cc_region_at(cc, i)  -- C extension is 1-based here too
  end
  return out
end

-- ---- ParsedNumber ----

--- A parsed phone number. Wraps the caller-owned parsed-number string the ABI
--- returns; its fields are read on demand through the pn_* accessors.
local ParsedNumber = {}
ParsedNumber.__index = ParsedNumber
M.ParsedNumber = ParsedNumber

function ParsedNumber.new(pn_string)
  return setmetatable({ _pn = pn_string }, ParsedNumber)
end

function ParsedNumber:region()
  return native.pn_region(self._pn)
end

function ParsedNumber:country_code()
  return native.pn_country_code(self._pn)
end

function ParsedNumber:national_number()
  return native.pn_national_number(self._pn)
end

function ParsedNumber:extension()
  return native.pn_extension(self._pn)
end

function ParsedNumber:italian_leading_zero()
  return native.pn_italian_leading_zero(self._pn)
end

function ParsedNumber:source()
  return native.pn_source(self._pn)
end

function ParsedNumber:error()
  return native.pn_error(self._pn)
end

function ParsedNumber:region_code()
  return native.region_code_for_number(self._pn)
end

function ParsedNumber:national_significant_number()
  return native.national_significant_number(self._pn)
end

function ParsedNumber:length_of_ndc()
  return native.length_of_ndc(self._pn)
end

function ParsedNumber:length_of_area_code()
  return native.length_of_area_code(self._pn)
end

function ParsedNumber:is_geographical()
  return native.is_geographical(self._pn)
end

--- Parse a human-typed number in a default region, returning a ParsedNumber.
function M.parse(input, region)
  return ParsedNumber.new(native.parse(input, region))
end

--- The national number extracted from raw input (cc + punctuation stripped).
function M.national_number(region, input)
  return native.national_number(region, input)
end

-- ---- validation ----

--- True if the national number is a length the region allows.
function M.is_possible_number(region, input)
  return native.is_possible_number(region, input)
end

--- Why (or that) a number is possible — a VR_* ValidationResult integer.
function M.is_possible_number_with_reason(region, input)
  return native.is_possible_number_with_reason(region, input)
end

--- True if the number matches the region's national-number patterns.
function M.is_valid_number(region, input)
  return native.is_valid_number(region, input)
end

--- True if a number is valid *for the given region* specifically.
function M.is_valid_number_for_region(input, region)
  return native.is_valid_number_for_region(input, region)
end

--- The number type (a TYPE_* integer; -1 for unknown).
function M.number_type(region, input)
  return native.number_type(region, input)
end

--- True if the number can be dialled from outside its country.
function M.can_be_internationally_dialled(region, input)
  return native.can_be_internationally_dialled(region, input)
end

-- ---- formatting ----

--- Format the number in the given style (E164 / INTERNATIONAL / NATIONAL /
--- RFC3966). `style` defaults to NATIONAL.
function M.format(region, input, style)
  return native.format(region, input, style or M.NATIONAL)
end

function M.format_national(region, input)
  return native.format(region, input, M.NATIONAL)
end

function M.format_international(region, input)
  return native.format(region, input, M.INTERNATIONAL)
end

function M.format_e164(region, input)
  return native.format(region, input, M.E164)
end

function M.format_rfc3966(region, input)
  return native.format(region, input, M.RFC3966)
end

--- Format as dialled from `calling_from` to a number in `region`.
function M.format_out_of_country(region, input, calling_from)
  return native.format_out_of_country(region, input, calling_from)
end

--- Format a ParsedNumber the way it was originally entered, from `calling_from`.
function M.format_in_original(parsed, calling_from)
  local pn = getmetatable(parsed) == ParsedNumber and parsed._pn or parsed
  return native.format_in_original(pn, calling_from)
end

-- ---- relations / helpers ----

--- Compare two numbers — a MATCH_* MatchType integer.
function M.is_number_match(a, b)
  return native.is_number_match(a, b)
end

--- Truncate an over-long number to the longest valid prefix, or "".
function M.truncate_too_long(region, input)
  return native.truncate_too_long(region, input)
end

--- Keep only the digits in a string (also mapping wide/fullwidth digits).
function M.normalize_digits_only(s)
  return native.normalize_digits_only(s)
end

--- Convert vanity letters to their dial digits (1-800-FLOWERS -> 1-800-3569377).
function M.convert_alpha_characters(s)
  return native.convert_alpha_characters(s)
end

--- True if the string contains vanity letters.
function M.is_alpha_number(s)
  return native.is_alpha_number(s)
end

-- ---- AsYouTypeFormatter ----

--- Formats a number as it is typed, digit by digit. The formatter state is a
--- caller-owned string the ABI hands back on each keypress; this object keeps
--- the current one and swaps it as characters arrive.
local AsYouTypeFormatter = {}
AsYouTypeFormatter.__index = AsYouTypeFormatter
M.AsYouTypeFormatter = AsYouTypeFormatter

function AsYouTypeFormatter.new(region)
  return setmetatable({ _state = native.ayt_new(region) }, AsYouTypeFormatter)
end

--- Feed one character; return the formatted-so-far string.
function AsYouTypeFormatter:input_digit(ch)
  self._state = native.ayt_input(self._state, tostring(ch))
  return self:result()
end

--- The formatted-so-far string, without feeding anything.
function AsYouTypeFormatter:result()
  return native.ayt_result(self._state)
end

--- Reset the formatter to empty.
function AsYouTypeFormatter:clear()
  self._state = native.ayt_clear(self._state)
end

-- ---- PhoneNumberMatcher / find_numbers ----

--- Find phone numbers in free text. Returns a list of matches, each a table
--- { start = <byte offset>, ["end"] = <byte offset>, raw = <matched text> }.
--- Offsets are the ABI's own (0-based) byte offsets into `text`.
function M.find_numbers(text, region, leniency)
  leniency = leniency or M.LENIENCY_VALID
  local out = {}
  local n = native.matcher_count(text, region, leniency)
  for i = 0, n - 1 do
    out[i + 1] = {
      start = native.matcher_start(text, region, leniency, i),
      ["end"] = native.matcher_end(text, region, leniency, i),
      raw = native.matcher_raw(text, region, leniency, i),
    }
  end
  return out
end

-- ---- ShortNumberInfo (short / emergency numbers) ----
-- Short numbers are dialled as-is (no cc, no national prefix): the input is the
-- raw short number plus a region.

--- True if `input` is a possible short number for the region (right length).
function M.short_is_possible(region, input)
  return native.short_is_possible(region, input)
end

--- True if `input` matches a short-number pattern for the region.
function M.short_is_valid(region, input)
  return native.short_is_valid(region, input)
end

--- True if `input` is an emergency number for the region (e.g. US "911").
function M.is_emergency_number(region, input)
  return native.short_is_emergency(region, input)
end

--- True if dialling `input` connects to an emergency number for the region.
function M.connects_to_emergency_number(region, input)
  return native.short_connects_to_emergency(region, input)
end

--- True if the short number is specific to a single carrier.
function M.short_is_carrier_specific(region, input)
  return native.short_is_carrier_specific(region, input)
end

--- True if the short number is usable as an SMS service.
function M.short_is_sms_service(region, input)
  return native.short_is_sms_service(region, input)
end

--- The expected cost of dialling the short number (a COST_* integer).
function M.short_expected_cost(region, input)
  return native.short_expected_cost(region, input)
end

--- An example short number for the region, or "".
function M.short_example_number(region)
  return native.short_example_number(region)
end

-- ---- PhoneNumberToTimeZonesMapper (timezone lookup) ----
-- Longest-prefix match over the number's E.164 digits: pass a raw (region,
-- input) like everywhere else and the engine parses to E.164 itself. The
-- unknown-zone sentinel is "Etc/Unknown".

--- The unknown-timezone sentinel, "Etc/Unknown".
function M.unknown_time_zone()
  return native.tz_unknown()
end

--- How many timezones the number maps to (0 = only the unknown zone).
function M.time_zone_count(region, input)
  return native.tz_count(region, input)
end

--- The IANA timezone ids for a number, as a list (a 1-based table). A number
--- with no known zones comes back as a single-element list holding the unknown
--- zone, never empty — matching the other bindings.
function M.time_zones_for_number(region, input)
  local n = native.tz_count(region, input)
  if n == 0 then
    return { M.unknown_time_zone() }
  end
  local out = {}
  for i = 0, n - 1 do        -- ABI idx is 0-based; build a 1-based list
    out[i + 1] = native.tz_at(region, input, i)
  end
  return out
end

-- ---- PhoneNumberToCarrierMapper (English carrier names) ----
-- Longest-prefix match over the E.164 digits; English names only. "" when no
-- carrier is known for the number.

--- The carrier name for a number (English), or "" if none is known.
function M.carrier_name_for_number(region, input)
  return native.carrier_name(region, input)
end

--- The carrier name only when the number is valid, else "".
function M.carrier_name_for_valid_number(region, input)
  return native.carrier_name_for_valid(region, input)
end

-- ---- lifecycle ----

--- The engine's ABI revision.
function M.abi_version()
  return native.abi_version()
end

--- Where the engine `.so` was actually loaded from.
function M.engine_path()
  return native.engine_path()
end

--- Force the engine to load from an explicit path (before first use). Returns
--- the resolved path.
function M.load(path)
  return native.load(path)
end

return M

File: elixir/lib/phonenumber_ae.ex
defmodule PhonenumberAe do
  @moduledoc """
  Validate, parse and format international phone numbers (ABI v2).

  This is a thin Elixir surface over the monorepo's **canonical BEAM NIF**,
  which lives in `erlang/` and is compiled exactly once. There is no C source
  in this directory and no second `.so` — every function here delegates to
  `:phonenumber_ae_nif`, the very same compiled module the Erlang and Gleam
  bindings load. One engine, one NIF, three languages.

  The engine itself (`core/native/libphonenumber_ae.so`) is pure Aether,
  compiled from Google libphonenumber's own metadata. No phone-number logic
  lives in this file: everything marshals to an `aether_pn_embed_*` call across
  the C ABI in `core/embed.ae` (docs/abi.md — 50 symbols, full
  PhoneNumberUtil parity).

      iex> PhonenumberAe.country_code("US")
      "1"
      iex> PhonenumberAe.is_valid_number?("US", "+1 201 555 0123")
      true
      iex> PhonenumberAe.format("US", "2015550123", :national)
      "(201) 555-0123"

  ## Finding the NIF

  `mix` does **not** honour `ERL_LIBS`, unlike `erl`, `escript` and `gleam`.
  So the app built by `erlang/.build.ae` is put on the code path explicitly —
  see `test/test_helper.exs`, which reads `$LIBPHONENUMBER_AE_BEAM_APP` and
  calls `Code.append_path/1`. That is the one wrinkle of consuming an OTP app
  that Mix did not build.

  ## No handles

  The ABI has no opaque handles. A parsed number and an as-you-type state are
  themselves caller-owned STRINGS: `PhonenumberAe.ParsedNumber` and
  `PhonenumberAe.AsYouTypeFormatter` wrap those strings, but every call is a
  direct FFI crossing.
  """

  @typedoc "A format style. Maps onto the ABI's integer selectors (append-only)."
  @type format_style :: :e164 | :international | :national | :rfc3966

  @typedoc "A PhoneNumberType. `:unknown` is the -1 sentinel."
  @type number_type ::
          :unknown
          | :fixed_line
          | :mobile
          | :toll_free
          | :premium_rate
          | :shared_cost
          | :voip
          | :personal_number
          | :pager
          | :uan
          | :voicemail

  @typedoc "A ValidationResult from `is_possible_number_with_reason/2`."
  @type validation_result ::
          :is_possible
          | :is_possible_local_only
          | :invalid_country_code
          | :too_short
          | :invalid_length
          | :too_long

  @typedoc "A MatchType from `is_number_match/2`."
  @type match_type :: :not_a_number | :no_match | :short_nsn | :nsn | :exact

  @typedoc "A CountryCodeSource, from a parsed number's `source`."
  @type country_code_source ::
          :from_number_with_plus
          | :from_number_with_idd
          | :from_number_without_plus
          | :from_default_country
          | :unspecified

  @typedoc "Matcher leniency."
  @type leniency :: :possible | :valid

  # ---- metadata ----

  @doc ~S'The country calling code for a region ("1", "44", …), or "" if unknown.'
  @spec country_code(iodata()) :: binary()
  defdelegate country_code(region), to: :phonenumber_ae_nif

  @doc ~S'An example national number for the region, or "".'
  @spec example_number(iodata()) :: binary()
  defdelegate example_number(region), to: :phonenumber_ae_nif

  @doc ~S'An example national number of a given type (`:mobile`, …), or "".'
  @spec example_number_for_type(iodata(), number_type() | integer()) :: binary()
  def example_number_for_type(region, type) when is_atom(type),
    do: :phonenumber_ae_nif.example_number_for_type(region, type_code(type))

  def example_number_for_type(region, type),
    do: :phonenumber_ae_nif.example_number_for_type(region, type)

  @doc ~S'An example number that is invalid for the region, or "".'
  @spec invalid_example_number(iodata()) :: binary()
  defdelegate invalid_example_number(region), to: :phonenumber_ae_nif

  @doc ~S'The possible-lengths spec for the region (e.g. "9,10"), or "".'
  @spec possible_lengths(iodata()) :: binary()
  defdelegate possible_lengths(region), to: :phonenumber_ae_nif

  @doc ~S'The (main) region for a country calling code, e.g. "44" -> "GB".'
  @spec region_code_for_country_code(iodata() | integer()) :: binary()
  def region_code_for_country_code(cc),
    do: :phonenumber_ae_nif.region_code_for_country_code(to_iodata(cc))

  @doc "True if the region is part of the North American Numbering Plan."
  @spec is_nanpa_country?(iodata()) :: boolean()
  def is_nanpa_country?(region),
    do: :phonenumber_ae_nif.is_nanpa_country(region) != 0

  @doc "The national-direct-dialling prefix for the region."
  @spec ndd_prefix_for_region(iodata(), boolean()) :: binary()
  def ndd_prefix_for_region(region, strip_non_digits \\ false),
    do: :phonenumber_ae_nif.ndd_prefix_for_region(region, bool_int(strip_non_digits))

  # ---- region enumeration ----

  @doc "How many regions the metadata carries."
  @spec region_count() :: non_neg_integer()
  defdelegate region_count(), to: :phonenumber_ae_nif

  @doc "The region id at `index` (0-based)."
  @spec region_at(integer()) :: binary()
  defdelegate region_at(index), to: :phonenumber_ae_nif

  @doc "Every region id the metadata carries, as a list of ISO-3166 codes."
  @spec regions() :: [binary()]
  defdelegate regions(), to: :phonenumber_ae_nif

  @doc "How many regions share a country calling code."
  @spec cc_region_count(iodata() | integer()) :: non_neg_integer()
  def cc_region_count(cc), do: :phonenumber_ae_nif.cc_region_count(to_iodata(cc))

  @doc "The region id at `index` among those sharing a country calling code."
  @spec cc_region_at(iodata() | integer(), integer()) :: binary()
  def cc_region_at(cc, index),
    do: :phonenumber_ae_nif.cc_region_at(to_iodata(cc), index)

  @doc ~S'Every region sharing a country calling code, e.g. "1" -> ["US", …].'
  @spec regions_for_country_code(iodata() | integer()) :: [binary()]
  def regions_for_country_code(cc) do
    bin = to_iodata(cc)
    n = :phonenumber_ae_nif.cc_region_count(bin)

    case n do
      0 -> []
      _ -> Enum.map(0..(n - 1), &:phonenumber_ae_nif.cc_region_at(bin, &1))
    end
  end

  # ---- parse ----

  @doc """
  Parse raw input into a `PhonenumberAe.ParsedNumber`. Read its fields with the
  struct's accessors; `error/1` is non-empty when the parse failed.
  """
  @spec parse(iodata(), iodata()) :: PhonenumberAe.ParsedNumber.t()
  def parse(input, region),
    do: PhonenumberAe.ParsedNumber.new(:phonenumber_ae_nif.parse(input, region))

  @doc "The national number extracted from raw input (cc + punctuation stripped)."
  @spec national_number(iodata(), iodata()) :: binary()
  defdelegate national_number(region, input), to: :phonenumber_ae_nif

  # ---- validity and typing ----

  @doc "True if the national number is a length the region allows."
  @spec is_possible_number?(iodata(), iodata()) :: boolean()
  def is_possible_number?(region, input),
    do: :phonenumber_ae_nif.is_possible_number(region, input) != 0

  @doc "The `t:validation_result/0` for the number (`:too_short`, …)."
  @spec is_possible_number_with_reason(iodata(), iodata()) :: validation_result()
  def is_possible_number_with_reason(region, input),
    do: validation_atom(:phonenumber_ae_nif.is_possible_number_with_reason(region, input))

  @doc "True if the number matches the region's national-number patterns."
  @spec is_valid_number?(iodata(), iodata()) :: boolean()
  def is_valid_number?(region, input),
    do: :phonenumber_ae_nif.is_valid_number(region, input) != 0

  @doc "True if the number is valid for the specific region (not just its cc)."
  @spec is_valid_number_for_region?(iodata(), iodata()) :: boolean()
  def is_valid_number_for_region?(input, region),
    do: :phonenumber_ae_nif.is_valid_number_for_region(input, region) != 0

  @doc "The `t:number_type/0` for the number (`:fixed_line`, `:mobile`, …)."
  @spec number_type(iodata(), iodata()) :: number_type()
  def number_type(region, input),
    do: type_atom(:phonenumber_ae_nif.number_type(region, input))

  @doc "The raw ABI type code (-1 unknown, 0 fixed_line, …)."
  @spec number_type_code(iodata(), iodata()) :: integer()
  def number_type_code(region, input),
    do: :phonenumber_ae_nif.number_type(region, input)

  @doc "True if the number can be dialled from outside its country."
  @spec can_be_internationally_dialled?(iodata(), iodata()) :: boolean()
  def can_be_internationally_dialled?(region, input),
    do: :phonenumber_ae_nif.can_be_internationally_dialled(region, input) != 0

  # ---- formatting ----
  #
  # The ABI takes an integer style; Elixir callers name it with an atom and
  # never see the number. These constants are ABI — append only, never
  # renumber (core/embed.ae). Note the v2 numbering: :e164 is 0 (was 2 in v1).

  @styles %{e164: 0, international: 1, national: 2, rfc3966: 3}

  @doc "Format the number in the given style."
  @spec format(iodata(), iodata(), format_style()) :: binary()
  def format(region, input, style \\ :national) when is_map_key(@styles, style),
    do: :phonenumber_ae_nif.format(region, input, Map.fetch!(@styles, style))

  @doc "`format/3` with the `:national` style."
  @spec format_national(iodata(), iodata()) :: binary()
  def format_national(region, input), do: format(region, input, :national)

  @doc "`format/3` with the `:international` style."
  @spec format_international(iodata(), iodata()) :: binary()
  def format_international(region, input), do: format(region, input, :international)

  @doc "`format/3` with the `:e164` style."
  @spec format_e164(iodata(), iodata()) :: binary()
  def format_e164(region, input), do: format(region, input, :e164)

  @doc "`format/3` with the `:rfc3966` style."
  @spec format_rfc3966(iodata(), iodata()) :: binary()
  def format_rfc3966(region, input), do: format(region, input, :rfc3966)

  @doc "Format as it would be dialled from `calling_from` (the region dialling out)."
  @spec format_out_of_country(iodata(), iodata(), iodata()) :: binary()
  defdelegate format_out_of_country(region, input, calling_from), to: :phonenumber_ae_nif

  @doc "Format a parsed number the way it was originally dialled."
  @spec format_in_original(PhonenumberAe.ParsedNumber.t(), iodata()) :: binary()
  def format_in_original(%PhonenumberAe.ParsedNumber{pn: pn}, calling_from),
    do: :phonenumber_ae_nif.format_in_original(pn, calling_from)

  # ---- relations / helpers ----

  @doc "Compare two numbers; returns a `t:match_type/0` (`:exact`, `:no_match`, …)."
  @spec is_number_match(iodata(), iodata()) :: match_type()
  def is_number_match(a, b),
    do: match_atom(:phonenumber_ae_nif.is_number_match(a, b))

  @doc "Drop digits past the region's maximum possible length."
  @spec truncate_too_long(iodata(), iodata()) :: binary()
  defdelegate truncate_too_long(region, input), to: :phonenumber_ae_nif

  @doc "Keep only the digits of a string (Unicode digits normalised to 0-9)."
  @spec normalize_digits_only(iodata()) :: binary()
  defdelegate normalize_digits_only(s), to: :phonenumber_ae_nif

  @doc "Convert vanity letters to their dial-pad digits (keeping punctuation)."
  @spec convert_alpha_characters(iodata()) :: binary()
  defdelegate convert_alpha_characters(s), to: :phonenumber_ae_nif

  @doc "True if the string contains any vanity letters."
  @spec is_alpha_number?(iodata()) :: boolean()
  def is_alpha_number?(s), do: :phonenumber_ae_nif.is_alpha_number(s) != 0

  # ---- find numbers ----

  @doc """
  Find phone numbers in free text (default leniency `:valid`). Returns a list of
  `PhonenumberAe.Match` structs, each carrying `start`, `end`, and `raw`.
  """
  @spec find_numbers(iodata(), iodata(), leniency()) :: [PhonenumberAe.Match.t()]
  def find_numbers(text, region, leniency \\ :valid) do
    l = leniency_code(leniency)
    n = :phonenumber_ae_nif.matcher_count(text, region, l)

    case n do
      0 ->
        []

      _ ->
        Enum.map(0..(n - 1), fn i ->
          %PhonenumberAe.Match{
            start: :phonenumber_ae_nif.matcher_start(text, region, l, i),
            end: :phonenumber_ae_nif.matcher_end(text, region, l, i),
            raw: :phonenumber_ae_nif.matcher_raw(text, region, l, i)
          }
        end)
    end
  end

  # ---- introspection ----

  @doc "The engine's ABI revision (2)."
  @spec abi_version() :: non_neg_integer()
  defdelegate abi_version(), to: :phonenumber_ae_nif

  # ---- internal ----

  defp bool_int(true), do: 1
  defp bool_int(false), do: 0

  defp to_iodata(cc) when is_integer(cc), do: Integer.to_string(cc)
  defp to_iodata(cc), do: cc

  # ABI number-type codes -> atoms. Append only, never renumber (core/embed.ae).
  defp type_atom(-1), do: :unknown
  defp type_atom(0), do: :fixed_line
  defp type_atom(1), do: :mobile
  defp type_atom(2), do: :toll_free
  defp type_atom(3), do: :premium_rate
  defp type_atom(4), do: :shared_cost
  defp type_atom(5), do: :voip
  defp type_atom(6), do: :personal_number
  defp type_atom(7), do: :pager
  defp type_atom(8), do: :uan
  defp type_atom(9), do: :voicemail
  # A newer engine could return an unseen code; degrade rather than crash.
  defp type_atom(_), do: :unknown

  defp type_code(:unknown), do: -1
  defp type_code(:fixed_line), do: 0
  defp type_code(:mobile), do: 1
  defp type_code(:toll_free), do: 2
  defp type_code(:premium_rate), do: 3
  defp type_code(:shared_cost), do: 4
  defp type_code(:voip), do: 5
  defp type_code(:personal_number), do: 6
  defp type_code(:pager), do: 7
  defp type_code(:uan), do: 8
  defp type_code(:voicemail), do: 9

  # ValidationResult codes -> atoms.
  defp validation_atom(0), do: :is_possible
  defp validation_atom(4), do: :is_possible_local_only
  defp validation_atom(1), do: :invalid_country_code
  defp validation_atom(2), do: :too_short
  defp validation_atom(5), do: :invalid_length
  defp validation_atom(3), do: :too_long
  defp validation_atom(_), do: :invalid_length

  # MatchType codes -> atoms.
  defp match_atom(0), do: :not_a_number
  defp match_atom(1), do: :no_match
  defp match_atom(2), do: :short_nsn
  defp match_atom(3), do: :nsn
  defp match_atom(4), do: :exact
  defp match_atom(_), do: :no_match

  # Matcher leniency codes.
  defp leniency_code(:possible), do: 0
  defp leniency_code(:valid), do: 1
end
